### PR TITLE
Optional blocks in Markdown cards

### PR DESCRIPTION
**Pull request by noahmoss**
Epic #23543

This PR adds support for using `[[ ]]` syntax to delineate part of a text card which should be optional, and only displayed when at least one parameter within it has a value.

For example, you could create a text card with the contents `Orders [[in {{state}}]` and connect it to a `state` filter. When the filter has no values,  the card would just display `Orders`.

I went down a rabbit hole trying to get some escaping-related edge cases working... for example `\[[{{foo}}]]` should not be parsed as an optional block, since the first square bracket is escaped. I ran into issues getting the regexes for this to work with the Closure compiler so I've omitted them from the PR. I don't think these are blocking issues in the slightest.

### CODE REVIEW

File: shared/src/metabase/shared/parameters/parameters.cljc [hunk 1]
```diff
@@ -113,36 +114,91 @@
...
       (try (-> (formatted-value tyype value locale)
                escape-chars)
            (catch #?(:clj Throwable :cljs js/Error) _
              ;; If we got an exception (most likely during date parsing/formatting), fallback to the default
              ;; implementation of formatted-value
-             (formatted-value :default value locale)))
-      ;; If this parameter has no value, return the original {{tag}} so that no substitution is done.
-      (first match))))
-
-(defn- normalize-parameter
```
Comment (noahmoss) [inline shared/src/metabase/shared/parameters/parameters.cljc:132]: This function I just moved down towards the bottom of the file

File: shared/src/metabase/shared/parameters/parameters.cljc [hunk 2]
```diff
@@ -113,36 +114,92 @@
...
 (def ^:private template-tag-regex
   "A regex to find template tags in a text card on a dashboard. This should mirror the regex used to find template
   tags in native queries, with the exception of snippets and card ID references (see the metabase-lib function
-  `recognizeTemplateTags` for that regex)."
+  `recognizeTemplateTags` for that regex).
+
+  If you modify this, also modify `template-tag-splitting-regex` below."
   #"\{\{\s*([A-Za-z0-9_\.]+?)\s*\}\}")
 
+; Represents a variable parsed out of a text card. `tag` contains the tag name alone, as a string. `source` contains
```
Comment (qnkhuat) [inline shared/src/metabase/shared/parameters/parameters.cljc:138]:
```suggestion
;; Represents a variable parsed out of a text card. `tag` contains the tag name alone, as a string. `source` contains
```

**Review by qnkhuat — APPROVED** [review]
👍 Only have a comment on the variable name.

File: shared/src/metabase/shared/parameters/parameters.cljc [hunk 3]
```diff
@@ -113,36 +114,92 @@
...
+; the full original syntax for the parameter)
+(defrecord ^:private TextVariable [tag source]
```
[the remark this answers is not in the document]
Comment (noahmoss) [reply shared/src/metabase/shared/parameters/parameters.cljc:140]: Fair, I was trying to distinguish it from the actual "parameter" data structures on dashboards, but maybe this makes it more confusing. I'll change it to `TextParam`

**Pull request merged**